File: src/test/java/legacy/DiscountCalculatorTest.java
package legacy;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static legacy.TestUtils.pln;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class DiscountCalculatorTest {

    private DiscountCalculator discountCalculator = new DiscountCalculator();

    @BeforeEach
    void setup() {
        discountCalculator = new DiscountCalculator();
    }

    @Test
    void shouldCalculateDiscount() {
        // given
        DiscountStrategy discountStrategy = new DiscountStrategy.Fixed(pln(100));

        // when
        Money result = discountCalculator.calculateDiscount(pln(150), discountStrategy);

        // then
        assertThat(result).isEqualTo(pln(50));
    }

    @Test
    void shouldLogInfo() {
        Logger loggerMock = Mockito.mock(Logger.class);
        discountCalculator = new DiscountCalculator(loggerMock);

        // given
        DiscountStrategy discountStrategy = new DiscountStrategy.Fixed(pln(100));

        // when
        discountCalculator.calculateDiscount(pln(10), discountStrategy);

        // then
        verify(loggerMock, times(1)).logInfo(anyString());
    }
}

File: src/test/kotlin/refactored/DiscountCalculatorTest.kt
package refactored


import assertk.assertThat
import assertk.assertions.isEqualTo
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class DiscountCalculatorTest {

    private var discountCalculator = DiscountCalculator()

    @BeforeEach
    fun setup() {
        discountCalculator = DiscountCalculator()
    }

    @Test
    fun `should calculate discount`() {
        // given
        val discountStrategy = DiscountStrategy.Fixed(pln(100))

        // when
        val result = discountCalculator.calculateDiscount(pln(150), discountStrategy)

        // then
        assertThat(result).isEqualTo(pln(50))
    }

    @Test
    fun `should log info`() {
        val loggerMock: Logger = mockk(relaxed = true)
        discountCalculator = DiscountCalculator(loggerMock)

        // given
        val discountStrategy = DiscountStrategy.Fixed(pln(100))

        // when
        discountCalculator.calculateDiscount(pln(10), discountStrategy)

        // then
        verify(exactly = 1) { loggerMock.logInfo(any()) }
    }
}